mod-pubsub-server/src/main/java/org/folio/services/impl/SecurityManagerImpl.java

```diff
@@ -61,33 +66,107 @@ public SecurityManagerImpl(@Autowired Vertx vertx, @Autowired Cache cache,
+    if (!StringUtils.isEmpty(cachedAccessToken)) {
+      LOGGER.debug("getAccessToken:: using cached access token for tenant {}",
+        params.getTenantId());
+      return Future.succeededFuture(cachedAccessToken);
     }
 
+    return logInWithExpiry(params)
+      .map(v -> cache.getAccessToken(tenantId));
+  }
+
+  public Future<Void> logInWithExpiry(OkapiConnectionParams params) {
+    final String tenantId = params.getTenantId();
+    LOGGER.info("logInWithExpiry:: Logging in, tenantId={}", tenantId);
+
     return Future.succeededFuture(systemUserConfig.getUserCredentialsJson())
-      .compose(userCredentials -> doRequest(params, LOGIN_URL, HttpMethod.POST, userCredentials.encode()))
+      .compose(userCredentials -> doRequest(params, LOGIN_WITH_EXPIRY_URL,
+        HttpMethod.POST, userCredentials.encode()))
       .compose(response -> {
         if (response.getCode() == HttpStatus.HTTP_CREATED.toInt()) {
-          LOGGER.info("Logged in {} user", systemUserConfig.getName());
-          String token = response.getResponse().getHeader(OKAPI_TOKEN_HEADER);
-          cache.addToken(params.getTenantId(), token);
-          return Future.succeededFuture(token);
+          LOGGER.info("logInWithExpiry:: Logged in {} user", systemUserConfig.getName());
+
+          ExpiryAwareToken accessToken = fetchTokenFromCookies(response, ACCESS_TOKEN_NAME, params);
+          if (accessToken == null) {
+            return tokenFetchFailure(ACCESS_TOKEN_NAME);
+          }
+
+          ExpiryAwareToken refreshToken = fetchTokenFromCookies(response, REFRESH_TOKEN_NAME,
+            params);
+          if (refreshToken == null) {
+            return tokenFetchFailure(REFRESH_TOKEN_NAME);
+          }
+
+          LOGGER.info("logInWithExpiry:: parsed access token and refresh token, caching");
+          cache.setAccessToken(tenantId, accessToken);
+          cache.setRefreshToken(tenantId, refreshToken);
+          return Future.succeededFuture();
         }
         String message = String.format("%s user was not logged in, received status %d",
-            systemUserConfig.getName(), response.getCode());
-        LOGGER.error(message);
+          systemUserConfig.getName(), response.getCode());
+        LOGGER.warn("logInWithExpiry:: {}", message);
         return Future.failedFuture(message);
       });
   }
 
+  public Future<Void> logInWithExpiry(ExpiryAwareToken token) {
+    return logInWithExpiry(token.getOkapiParams());
+  }
+
+  private ExpiryAwareToken fetchTokenFromCookies(RestUtil.WrappedResponse response,
+    String tokenName, OkapiConnectionParams params) {
+
+    String tokenCookie = response.getResponse().cookies().stream()
+      .filter(t -> t.contains(tokenName))
+      .findFirst()
+      .orElse(null);
+
+    if (tokenCookie != null) {
+      String token = Arrays.stream(tokenCookie.split(";"))
+        .map(String::trim)
+        .filter(param -> param.startsWith(tokenName))
+        .map(param -> param.split("="))
```

Comment:
Should we do it the other way around and match more strictly?
```suggestion
        .map(param -> param.split("="))
        .filter(param -> tokenName.equals(param))

```